Optimize random

chris0x44 opened the pull request:
# Description
Applied the multiply-shift optimization described by Daniel Lemire (https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/ which also refers to arXiv:1805.10941v4).

Note that the the bias-check was changed. The `_Ret/_Index` calculation seemed to be optimized in a way that it could easily return `_Ret%_Index`, making the modulo itself relatively cheap. For a sample testcase, which I compared to the original implementation as baseline, I found that only applying multiply-shift is around 1.1 times faster. Changing the check to `_Ret < (_Mask - _Mask % _Index)` in addition to the optimization showed to be 1.5 times faster. To be safe I also tried keeping the modulo operation and used just the modified bias-check, which was 1.2 only times faster than the baseline.

In the spirit of removing modulo operations, I also replaced the check `_Mask % _Index == _Udiff(_Index - 1)` (if _Index is a power of 2) with a check that does not rely on modulo operations.

Fixes #178.

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [ ] These changes were written from scratch using only this repository and
  the C++ Working Draft as a reference (and any other cited standards).
  If they were derived from a project that's already listed in NOTICE.txt,
  that's fine, but please mention it. If they were derived from any other
  project (including Boost and libc++, which are not yet listed in
  NOTICE.txt), you *must* mention it here, so we can determine whether the
  license is compatible and what else needs to be done.

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4889,23 +4889,120 @@ public:
+        return (__hi << (64 - __dist)) | (__lo >> __dist);
+#else // ^^^ 64-bit ^^^ / vvv 32-bit vvv
+      // Avoid a 64-bit shift by taking advantage of the range of shift values.
+        _STL_INTERNAL_CHECK(__dist >= 32);
```
This code was copied from charconv/Ryu (as previous comments indicate). This raises a couple of issues which need to be addressed: (1) as this comment indicates, there is a Ryu-specific assumption about the range of possible shift values which is probably not applicable here, and (2) if we copy Ryu code around in our repo, we need to ensure that the Boost license banner with Ulf Adams' copyright notice appears in those files.

(This also means that you should test both x86 and x64.)

chris0x44 replied:
Yes, I missed the range restriction in the x86 case. That need's to be adapted.

Regarding (2), I'm not sure how to correctly achieve this. Would it "just" be adding the copyright section of xcharconv_ryu.h (lines 7-32) to xutility? Perhaps as an alternative, I'd ask if copying the code is a good idea in the first place or if a separate header providing a ´_Umul128´ function is preferable, which would also make attribution more clear.

StephanTLavavej replied:
I'm increasingly liking your idea of a separate header for the attribution reason you've mentioned. We generally try to avoid having lots of tiny headers, but it's not an ironclad rule, and we could even avoid code duplication (since the umul128 doesn't need to be modified).

For now, what you have is fine. I'll need to think about how to rearrange this with minimal disruption to charconv (where https://github.com/StephanTLavavej/ryu/tree/msvc-latest is layering a massive number of changes on top of the upstream sources); when I figure that out, I'll want to push the change to your branch. I'll leave this active as a reminder to myself.

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4889,23 +4889,120 @@ public:
+
+#endif // ^^^ intrinsics unavailable ^^^
+
+    // using multiply-shift instead of modulo to fairly map values in range [0, _Mod), adapted from arXiv:1805.10941v4
```
We'll need to check if this is OK. Did you refer to only the blog post and the PDF on arXiv, but not the `fastrange` repository?

chris0x44 replied:
Well, that's embarrassing. I actually missed the github link to the standalone header, as I figured that would also refer to the file linked to from "As usual, my code is freely available."

StephanTLavavej replied:
Thanks; I'm looking into this now.

StephanTLavavej replied:
We're good to go! I'm busy with other things this week, but I should be able to update your PR soon.

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4889,23 +4889,120 @@ public:
+            return static_cast<_Udiff>((static_cast<uint64_t>(_Value) * _Mod) >> _Bitrange);
+        }
+
+        uint64_t __val_hi = 0;
```
We always use `_Leading_underscore_capital` for `_Ugly` identifiers, with the sole exception of Ryu-derived code in `charconv`, where `__leading_double_underscore_lowercase` was mechanically easier to search-and-replace (this was before VSCode gained case-preserving search-and-replace, but even with that, `__a` avoids the `_A` macro-like form that we always avoid).

chris0x44 replied:
OK, will change it to ´_Ugly´-style.

StephanTLavavej replied:
Let's leave the Ryu-derived code as-is; that will make it easier to spot differences if upstream changes.

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4889,23 +4889,120 @@ public:
+        } else if constexpr (_Udiff_bits == _Uint_bits) {
+            // ensure promotion to ull to keep all bits
+            return static_cast<_Udiff>((static_cast<uint64_t>(_Value) * _Mod) >> _Bitrange);
+        }
```
This should be followed by `else`. For ordinary control flow, we prefer to avoid chaining after unconditional `return`, which reduces the amount of nesting. But for `if constexpr`, the `else` prevents that codepath from being emitted at all when the previous condition(s) are selected.

chris0x44 replied:
OK, will be changed

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4889,23 +4889,120 @@ public:
+        // We don't need to handle the case __dist >= 64 here (see above).
+        _STL_INTERNAL_CHECK(__dist < 64);
+#ifdef _WIN64
+        _STL_INTERNAL_CHECK(__dist > 0);
```
Note that this assumption (nonzero shift) is a Ryu assumption; does your code guarantee that the shift is always nonzero?

chris0x44 replied:
Yes, since it will only be called to with _Bits as parameter. As _Bits stores the number of bits generated by urgn, which must provide at least one bit, it can not be zero. I take that from the requirement `G​::​min() < G​::​max()` of uniform_random_bit_generator.

StephanTLavavej commented:
/azp run

chris0x44 commented:
Reworked the code since I realized that I was reducing to the wrong number of bits. Initially I passed `_Bits` to `_Reduce` which was wrong for values created from multiple `_Get_bits` calls. In that case the number of bits should have been summed up.
This also required changes to  `_Reduce` where the choice of operation can now not be determined at compile-time anymore.

BillyONeal commented:
Sorry I've been relatively quiet on this. I touched random most recently but have been out of town a lot.

Do you have benchmark results between the old implementation and the new implementation to indicate how much faster it is?

chris0x44 commented:
@BillyONeal Yes, I did some benchmarks to catch regressions. The post-fix of the listed functions indicate the size of the (unsigned) return-type.


> Run on (8 X 2592 MHz CPU s)
> CPU Caches:
>   L1 Data 32K (x4)
>   L1 Instruction 32K (x4)
>   L2 Unified 262K (x4)
>   L3 Unified 6291K (x1)
> 

**Engine: std::mt19937**
Target: x86
```
-------------------------------------------------------
Benchmark             Time             CPU   Iterations
-------------------------------------------------------
baseline_64        34.9 ns         34.5 ns     19478261
optimized_64       25.4 ns         25.1 ns     28000000

baseline_32        14.0 ns         14.1 ns     49777778
optimized_32       11.0 ns         11.0 ns     64000000

baseline_16        14.2 ns         14.1 ns     49777778
optimized_16       10.8 ns         10.7 ns     64000000

baseline_8         13.8 ns         13.8 ns     49777778
optimized_8        11.0 ns         11.0 ns     64000000
```

Target: x64
```
-------------------------------------------------------
Benchmark             Time             CPU   Iterations
-------------------------------------------------------
baseline_64        31.1 ns         30.7 ns     22400000
optimized_64       17.2 ns         16.9 ns     40727273

baseline_32        13.5 ns         13.5 ns     49777778
optimized_32       8.80 ns         8.79 ns     74666667

baseline_16        13.8 ns         13.7 ns     56000000
optimized_16       8.75 ns         8.79 ns     74666667

baseline_8         12.9 ns         12.8 ns     56000000
optimized_8        8.50 ns         8.54 ns     89600000
```


**Engine: std::ranlux24_base**
Target: x86
```
baseline_64        65.8 ns         65.6 ns     11200000
optimized_64       35.6 ns         35.9 ns     21333333

baseline_32        22.5 ns         22.5 ns     26352941
optimized_32       19.7 ns         20.0 ns     32000000

baseline_16        15.6 ns         15.7 ns     44800000
optimized_16       12.5 ns         12.3 ns     56000000

baseline_8         22.3 ns         22.0 ns     29866667
optimized_8        11.8 ns         11.7 ns     56000000
```

Target: x64
```
baseline_64        39.2 ns         39.1 ns     16000000
optimized_64       21.3 ns         21.5 ns     32000000

baseline_32        20.5 ns         19.9 ns     29866667
optimized_32       13.2 ns         13.3 ns     44800000

baseline_16        13.1 ns         13.1 ns     56000000
optimized_16       9.01 ns         9.00 ns     74666667

baseline_8         13.0 ns         13.1 ns     56000000
optimized_8        8.63 ns         8.54 ns     89600000
```

CaseyCarter commented on stl/inc/xutility:
```diff
@@ -4889,24 +4917,121 @@ public:
+        return _Value && !(_Value & (_Value - 1));
+    }
+
+    _Diff operator()(_Diff _Index) { // adapt _Urng closed range to [0, _Index)
```
`_NODISCARD` seems appropriate here.

chris0x44 replied:
will add

BillyONeal commented:
Thanks for the benchmark data!

BillyONeal commented:
Unfortunately this results in a fair number of test failures, I think we need to get the tests out here on GH before we'll want to merge this one. We can't fault you for tests you can't see!

```
failure: CPPTEST-56-03 devcrt tests\CPP\Dev09_181509_tr1_inf_loop_uniform_int_ull /nologo /Od /W4 /w14061 /w14242 /w14582 /w14583 /w14587 /w14588 /w14265 /w14365 /w14749 /w14841 /w14842 /w15038 /sdl /WX /Zc:strictStrings /D_ENABLE_STL_INTERNAL_CHECK /D_ENFORCE_FACET_SPECIALIZATIONS=1 /bigobj /EHsc /MD /D_ITERATOR_DEBUG_LEVEL=0 /std:c++17 1
###   Test executable: D:\Contest\xnn4cvld.333\src\tools\perl\bin\perl.exe
###    with arguments: -S "D:\Contest\xnn4cvld.333\src\qa\vc\libs\devcrt\run.pl"
### Working directory: D:\Contest\xnn4cvld.333\src\qa\vc\libs\devcrt\tests\CPP\Dev09_181509_tr1_inf_loop_uniform_int_ull
### Test exited with code 0x00000003
### Test ran for 0:00:00:04.7735584
### Environment:
set CLANG_TARGET=-m32
set HAS_EDG=yes
set INCLUDE=D:\Contest\xnn4cvld.333\binaries\x86ret\inc;D:\Contest\xnn4cvld.333\src\ExternalAPIs\Windows\10\sdk\inc;D:\Contest\xnn4cvld.333\src\ExternalAPIs\UnifiedCRT\inc;D:\Contest\xnn4cvld.333\src\qa\vc\libs\devcrt\include
set LIB=D:\Contest\xnn4cvld.333\binaries\x86ret\lib\i386;D:\Contest\xnn4cvld.333\src\ExternalAPIs\Windows\10\sdk\lib\x86;D:\Contest\xnn4cvld.333\src\ExternalAPIs\UnifiedCRT\lib\i386;D:\Contest\xnn4cvld.333\src\ExternalAPIs\NetFX\v4.5\lib\x86;D:\Contest\xnn4cvld.333\src\vctools\nonship\cuda\v9.2\lib\x64
set PATH=D:\Contest\xnn4cvld.333\binaries\x86ret\bin\i386;D:\Contest\xnn4cvld.333\src\qa\vc\shared\testenv\bin;D:\Contest\xnn4cvld.333\src\tools\perl\bin;D:\Contest\xnn4cvld.333\src\vctools\nonship\clangllvm\x86\bin;D:\Contest\xnn4cvld.333\src\vctools\nonship\cuda\v9.2\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files\Dell\SysMgt\oma\bin;C:\Program Files\Dell\SysMgt\shared\bin;C:\Program Files\Dell\SysMgt\idrac;d:\git-lfs;C:\Users\VCWRKSPC\AppData\Local\Microsoft\WindowsApps
set PM_CL=/nologo /Od /W4 /w14061 /w14242 /w14582 /w14583 /w14587 /w14588 /w14265 /w14365 /w14749 /w14841 /w14842 /w15038 /sdl /WX /Zc:strictStrings /D_ENABLE_STL_INTERNAL_CHECK /D_ENFORCE_FACET_SPECIALIZATIONS=1 /bigobj /EHsc /MD /D_ITERATOR_DEBUG_LEVEL=0 /std:c++17
set TARGET_ARCHITECTURE=x86
### CLI output:
Executing: "D:\Contest\xnn4cvld.333\src\qa\vc\shared\testenv\bin\run.pl"
RUN.PL line 41:  _CONFIG_PREFIX  NOT SET
RUN.PL line 61:  Building..
RUN.PL line 90:  Using default build step
RUN.PM line 746:  Executing: "cl  /nologo /Od /W4 /w14061 /w14242 /w14582 /w14583 /w14587 /w14588 /w14265 /w14365 /w14749 /w14841 /w14842 /w15038 /sdl /WX /Zc:strictStrings /D_ENABLE_STL_INTERNAL_CHECK /D_ENFORCE_FACET_SPECIALIZATIONS=1 /bigobj /EHsc /MD /D_ITERATOR_DEBUG_LEVEL=0 /std:c++17   .\test.cpp   /c".
test.cpp
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4963): error C2220: the following warning is treated as an error
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4946): note: while compiling class template member function '_Diff std::_Rng_from_urng<_Diff,_Engine>::operator ()(_Diff)'
        with
        [
            _Diff=unsigned short,
            _Engine=Wacky
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1859): note: see reference to function template instantiation '_Diff std::_Rng_from_urng<_Diff,_Engine>::operator ()(_Diff)' being compiled
        with
        [
            _Diff=unsigned short,
            _Engine=Wacky
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1849): note: see reference to class template instantiation 'std::_Rng_from_urng<unsigned short,_Engine>' being compiled
        with
        [
            _Engine=Wacky
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1817): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::_Eval<_Engine>(_Engine &,_Ty,_Ty) const' being compiled
        with
        [
            _Ty=int16_t,
            _Engine=Wacky
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1817): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::_Eval<_Engine>(_Engine &,_Ty,_Ty) const' being compiled
        with
        [
            _Ty=int16_t,
            _Engine=Wacky
        ]
.\test.cpp(24): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::operator ()<URNG>(_Engine &) const' being compiled
        with
        [
            _Ty=int16_t,
            URNG=Wacky,
            _Engine=Wacky
        ]
.\test.cpp(24): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::operator ()<URNG>(_Engine &) const' being compiled
        with
        [
            _Ty=int16_t,
            URNG=Wacky,
            _Engine=Wacky
        ]
.\test.cpp(83): note: see reference to function template instantiation 'void microtest<URNG,int16_t,-32768,32767>(void)' being compiled
        with
        [
            URNG=Wacky
        ]
.\test.cpp(119): note: see reference to function template instantiation 'void add_tests<Wacky>(std::vector<fp_t,std::allocator<fp_t>> &)' being compiled
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4963): warning C4244: 'return': conversion from 'unsigned long' to '_Diff', possible loss of data
        with
        [
            _Diff=unsigned short
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4963): warning C4244: 'return': conversion from 'unsigned __int64' to '_Diff', possible loss of data
        with
        [
            _Diff=unsigned int
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4946): note: while compiling class template member function '_Diff std::_Rng_from_urng<_Diff,_Engine>::operator ()(_Diff)'
        with
        [
            _Diff=unsigned int,
            _Engine=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1859): note: see reference to function template instantiation '_Diff std::_Rng_from_urng<_Diff,_Engine>::operator ()(_Diff)' being compiled
        with
        [
            _Diff=unsigned int,
            _Engine=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1849): note: see reference to class template instantiation 'std::_Rng_from_urng<unsigned int,_Engine>' being compiled
        with
        [
            _Engine=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1817): note: see reference to function template instantiation 'int std::uniform_int<_Ty>::_Eval<_Engine>(_Engine &,_Ty,_Ty) const' being compiled
        with
        [
            _Ty=int32_t,
            _Engine=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1817): note: see reference to function template instantiation 'int std::uniform_int<_Ty>::_Eval<_Engine>(_Engine &,_Ty,_Ty) const' being compiled
        with
        [
            _Ty=int32_t,
            _Engine=std::mt19937_64
        ]
.\test.cpp(24): note: see reference to function template instantiation 'int std::uniform_int<_Ty>::operator ()<URNG>(_Engine &) const' being compiled
        with
        [
            _Ty=int32_t,
            URNG=std::mt19937_64,
            _Engine=std::mt19937_64
        ]
.\test.cpp(24): note: see reference to function template instantiation 'int std::uniform_int<_Ty>::operator ()<URNG>(_Engine &) const' being compiled
        with
        [
            _Ty=int32_t,
            URNG=std::mt19937_64,
            _Engine=std::mt19937_64
        ]
.\test.cpp(83): note: see reference to function template instantiation 'void microtest<URNG,int32_t,-2147483648,2147483647>(void)' being compiled
        with
        [
            URNG=std::mt19937_64
        ]
.\test.cpp(118): note: see reference to function template instantiation 'void add_tests<std::mt19937_64>(std::vector<fp_t,std::allocator<fp_t>> &)' being compiled
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4963): warning C4244: 'return': conversion from 'unsigned __int64' to '_Diff', possible loss of data
        with
        [
            _Diff=unsigned short
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4946): note: while compiling class template member function '_Diff std::_Rng_from_urng<_Diff,_Engine>::operator ()(_Diff)'
        with
        [
            _Diff=unsigned short,
            _Engine=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1859): note: see reference to function template instantiation '_Diff std::_Rng_from_urng<_Diff,_Engine>::operator ()(_Diff)' being compiled
        with
        [
            _Diff=unsigned short,
            _Engine=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1849): note: see reference to class template instantiation 'std::_Rng_from_urng<unsigned short,_Engine>' being compiled
        with
        [
            _Engine=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1817): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::_Eval<_Engine>(_Engine &,_Ty,_Ty) const' being compiled
        with
        [
            _Ty=int16_t,
            _Engine=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1817): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::_Eval<_Engine>(_Engine &,_Ty,_Ty) const' being compiled
        with
        [
            _Ty=int16_t,
            _Engine=std::mt19937_64
        ]
.\test.cpp(24): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::operator ()<URNG>(_Engine &) const' being compiled
        with
        [
            _Ty=int16_t,
            URNG=std::mt19937_64,
            _Engine=std::mt19937_64
        ]
.\test.cpp(24): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::operator ()<URNG>(_Engine &) const' being compiled
        with
        [
            _Ty=int16_t,
            URNG=std::mt19937_64,
            _Engine=std::mt19937_64
        ]
.\test.cpp(83): note: see reference to function template instantiation 'void microtest<URNG,int16_t,-32768,32767>(void)' being compiled
        with
        [
            URNG=std::mt19937_64
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4963): warning C4242: 'return': conversion from 'unsigned int' to '_Diff', possible loss of data
        with
        [
            _Diff=unsigned short
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\xutility(4946): note: while compiling class template member function '_Diff std::_Rng_from_urng<_Diff,_Engine>::operator ()(_Diff)'
        with
        [
            _Diff=unsigned short,
            _Engine=std::mt19937
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1859): note: see reference to function template instantiation '_Diff std::_Rng_from_urng<_Diff,_Engine>::operator ()(_Diff)' being compiled
        with
        [
            _Diff=unsigned short,
            _Engine=std::mt19937
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1849): note: see reference to class template instantiation 'std::_Rng_from_urng<unsigned short,_Engine>' being compiled
        with
        [
            _Engine=std::mt19937
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1817): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::_Eval<_Engine>(_Engine &,_Ty,_Ty) const' being compiled
        with
        [
            _Ty=int16_t,
            _Engine=std::mt19937
        ]
D:\Contest\xnn4cvld.333\binaries\x86ret\inc\random(1817): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::_Eval<_Engine>(_Engine &,_Ty,_Ty) const' being compiled
        with
        [
            _Ty=int16_t,
            _Engine=std::mt19937
        ]
.\test.cpp(24): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::operator ()<URNG>(_Engine &) const' being compiled
        with
        [
            _Ty=int16_t,
            URNG=std::mt19937,
            _Engine=std::mt19937
        ]
.\test.cpp(24): note: see reference to function template instantiation 'short std::uniform_int<_Ty>::operator ()<URNG>(_Engine &) const' being compiled
        with
        [
            _Ty=int16_t,
            URNG=std::mt19937,
            _Engine=std::mt19937
        ]
.\test.cpp(83): note: see reference to function template instantiation 'void microtest<URNG,int16_t,-32768,32767>(void)' being compiled
        with
        [
            URNG=std::mt19937
        ]
.\test.cpp(117): note: see reference to function template instantiation 'void add_tests<std::mt19937>(std::vector<fp_t,std::allocator<fp_t>> &)' being compiled
RUN.PM line 749:  Execution of: "cl  /nologo /Od /W4 /w14061 /w14242 /w14582 /w14583 /w14587 /w14588 /w14265 /w14365 /w14749 /w14841 /w14842 /w15038 /sdl /WX /Zc:strictStrings /D_ENABLE_STL_INTERNAL_CHECK /D_ENFORCE_FACET_SPECIALIZATIONS=1 /bigobj /EHsc /MD /D_ITERATOR_DEBUG_LEVEL=0 /std:c++17   .\test.cpp   /c" returned "2".
RUN.PM line 1040 - Error:  cl failed.  Cascading.
```
Example input:
```
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

#include <algorithm>
#include <assert.h>
#include <execution>
#include <random>
#include <stdint.h>
#include <vector>

using namespace std;
using namespace std::execution;

// REQUIRE: cpp17

const uint32_t N = 500'000;

template <typename URNG, typename T, T A, T B>
void microtest() {
    URNG engine;
    uniform_int_distribution<T> dist(A, B);

    for (uint32_t i = 0; i < N; ++i) {
        const T val = dist(engine);

        assert(val >= A && val <= B);
    }
}

using fp_t = void (*)();
template <typename URNG>
void add_tests(vector<fp_t>& tests) {
    // Test DevDiv#83370 "uniform_int_distribution isn't uniform".
    tests.insert(
        tests.end(), {
                         microtest<URNG, int16_t, INT16_MIN, INT16_MAX>,
                         microtest<URNG, int16_t, INT16_MIN + 1, INT16_MAX>,
                         microtest<URNG, int16_t, INT16_MIN, INT16_MAX - 1>,
                         microtest<URNG, int16_t, INT16_MIN + 2, INT16_MAX - 3>,
                         microtest<URNG, int16_t, -4, 4>,
                         microtest<URNG, int16_t, 11, 22>,
                         microtest<URNG, int16_t, -44, -33>,
                         microtest<URNG, int16_t, 5, 6>,
                         microtest<URNG, int16_t, -7, -7>,

                         microtest<URNG, uint16_t, 0, UINT16_MAX>,
                         microtest<URNG, uint16_t, 1, UINT16_MAX>,
                         microtest<URNG, uint16_t, 0, UINT16_MAX - 1>,
                         microtest<URNG, uint16_t, 2, UINT16_MAX - 3>,
                         microtest<URNG, uint16_t, 123, 456>,
                         microtest<URNG, uint16_t, 1000, 1001>,
                         microtest<URNG, uint16_t, 777, 777>,

                         microtest<URNG, int32_t, INT32_MIN, INT32_MAX>,
                         microtest<URNG, int32_t, INT32_MIN + 2, INT32_MAX - 3>,
                         microtest<URNG, uint32_t, 0, UINT32_MAX>,
                         microtest<URNG, uint32_t, 2, UINT32_MAX - 3>,
                         microtest<URNG, int64_t, INT64_MIN, INT64_MAX>,
                         microtest<URNG, int64_t, INT64_MIN + 2, INT64_MAX - 3>,
                         microtest<URNG, uint64_t, 0, UINT64_MAX>, // Test DDB#181509 "TR1 VC9 SP1: Infinite loop in
                                                                   // uniform_int<unsigned long long>::_Eval()".
                         microtest<URNG, uint64_t, 2, UINT64_MAX - 3>,

                         microtest<URNG, int32_t, -4, 4>,
                         microtest<URNG, int32_t, 11, 22>,
                         microtest<URNG, int32_t, -44, -33>,
                         microtest<URNG, int32_t, 5, 6>,
                         microtest<URNG, int32_t, -7, -7>,

                         microtest<URNG, uint32_t, 123, 456>,
                         microtest<URNG, uint32_t, 1000, 1001>,
                         microtest<URNG, uint32_t, 777, 777>,

                         microtest<URNG, int64_t, -4, 4>,
                         microtest<URNG, int64_t, 11, 22>,
                         microtest<URNG, int64_t, -44, -33>,
                         microtest<URNG, int64_t, 5, 6>,
                         microtest<URNG, int64_t, -7, -7>,

                         microtest<URNG, uint64_t, 123, 456>,
                         microtest<URNG, uint64_t, 1000, 1001>,
                         microtest<URNG, uint64_t, 777, 777>,
                     });
}

class Wacky {
public:
    Wacky() : mt(1701) {}

    typedef unsigned long result_type;

    static result_type min() {
        return 0;
    }

    static result_type max() {
        return 90000;
    }

    result_type operator()() {
        for (;;) {
            const result_type ret = mt() & 0x1FFFFUL;

            if (ret >= min() && ret <= max()) {
                return ret;
            }
        }
    }

private:
    mt19937 mt;
};

int main() {
    vector<fp_t> tests;

    add_tests<mt19937>(tests);
    add_tests<mt19937_64>(tests);
    add_tests<Wacky>(tests);

    for_each(par, tests.begin(), tests.end(), [](fp_t fn) { fn(); });
}
```

BillyONeal commented:
Here's another failure:

```
C:\Dev\msvc>.\src\tools\jobify\jobify.exe clang-cl /std:c++latest /Zi /EHsc /nologo /D_ENABLE_STL_INTERNAL_CHECK /MTd /Isrc\vctools\crt\github\stl\inc /Isrc\qa\vc\libs\devcrt\include .\src\qa\vc\libs\devcrt\tests\CPP\Dev09_192736_tr1_prngs_not_copyconstructible\test.cpp
[jobify] Usage: jobify.exe subcommand
[jobify] Executing: clang-cl /std:c++latest /Zi /EHsc /nologo /D_ENABLE_STL_INTERNAL_CHECK /MTd /Isrc\vctools\crt\github\stl\inc /Isrc\qa\vc\libs\devcrt\include .\src\qa\vc\libs\devcrt\tests\CPP\Dev09_192736_tr1_prngs_not_copyconstructible\test.cpp
[jobify] Command exited with 0

C:\Dev\msvc>.\test.exe
```
![image](https://user-images.githubusercontent.com/1544943/68986958-70ccf780-07d9-11ea-9af4-7e39a3cf9f73.png)

And the input:
```
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

#include <algorithm>
#include <assert.h>
#include <functional>
#include <iterator>
#include <random>
#include <vector>

using namespace std;

template <typename Gen>
void test_generator_can_be_value_direct_initialized() {
    Gen gen;

    Gen other(gen);

    assert(other == gen);
}

template <typename Gen>
void test_generator_can_be_const_ref_direct_initialized() {
    Gen gen;
    const Gen& r = gen;

    Gen other(r);

    assert(other == gen);
}

template <typename Gen>
void test_generator_can_be_value_copy_initialized() {
    Gen gen;

    Gen other = gen;

    assert(other == gen);
}

template <typename Gen>
void test_generator_can_be_const_ref_copy_initialized() {
    Gen gen;
    const Gen& r = gen;

    Gen other = r;

    assert(other == gen);
}

template <typename Gen, typename ReseedFunc>
void test_generator_can_be_reseeded(ReseedFunc reseedFunc) {
    const auto iterations = 100u;
    Gen gen;
    reseedFunc(gen);
    vector<typename Gen::result_type> firstPassResults;
    generate_n(back_inserter(firstPassResults), iterations, ref(gen));
    reseedFunc(gen);
    vector<typename Gen::result_type> seededResults;
    generate_n(back_inserter(seededResults), iterations, ref(gen));
    assert(firstPassResults == seededResults);
}

template <typename Gen>
void test_engine() {
    test_generator_can_be_value_direct_initialized<Gen>();
    test_generator_can_be_const_ref_direct_initialized<Gen>();
    test_generator_can_be_value_copy_initialized<Gen>();
    test_generator_can_be_const_ref_copy_initialized<Gen>();
    test_generator_can_be_reseeded<Gen>([](Gen& gen) { gen.seed(); });
    test_generator_can_be_reseeded<Gen>([](Gen& gen) { gen.seed(1729); });
}

template <typename Dist>
void test_distribution_can_be_reset(Dist dist) {
    knuth_b gen;
    const auto iterations = 100u;
    vector<typename Dist::result_type> firstPassResults;
    generate_n(back_inserter(firstPassResults), iterations, [&]() { return dist(gen); });
    dist.reset();
    gen.seed();
    vector<typename Dist::result_type> resetResults;
    generate_n(back_inserter(resetResults), iterations, [&]() { return dist(gen); });
    assert(firstPassResults == resetResults);
}

template <typename Dist>
void test_distribution(Dist dist) {
    test_distribution_can_be_reset(dist);
}

template <template <typename> class Dist>
void test_integral_distribution() {
    test_distribution(Dist<unsigned short>{});
    test_distribution(Dist<unsigned int>{});
    test_distribution(Dist<unsigned long>{});
    test_distribution(Dist<unsigned long long>{});
    test_distribution(Dist<short>{});
    test_distribution(Dist<int>{});
    test_distribution(Dist<long>{});
    test_distribution(Dist<long long>{});
}

template <template <typename> class Dist>
void test_real_distribution() {
    test_distribution(Dist<float>{});
    test_distribution(Dist<double>{});
    test_distribution(Dist<long double>{});
}

int main() {
    // N4527 26.5.5 [rand.predef], in paragraph order:
    test_engine<minstd_rand0>(); // /1
    test_engine<minstd_rand>();
    test_engine<mt19937>();
    test_engine<mt19937_64>();
    test_engine<ranlux24_base>(); // /5
    test_engine<ranlux48_base>();
    test_engine<ranlux24>();
    test_engine<ranlux48>();
    test_engine<knuth_b>();
    test_engine<default_random_engine>(); // /10

    // independent_bits_engine is not in any of the [rand.predef] engines, so
    // test it explicitly:
    test_engine<independent_bits_engine<mt19937_64, 5, unsigned int>>();
    // random_device can't be copied or reseeded, so no reason to test it here.

    test_integral_distribution<uniform_int_distribution>();
    test_real_distribution<uniform_real_distribution>();
    test_distribution(bernoulli_distribution{});
    test_integral_distribution<binomial_distribution>();
    test_integral_distribution<geometric_distribution>();
    test_integral_distribution<negative_binomial_distribution>();
    test_integral_distribution<poisson_distribution>();
    test_real_distribution<exponential_distribution>();
    test_real_distribution<gamma_distribution>();
    test_real_distribution<weibull_distribution>();
    test_real_distribution<extreme_value_distribution>();
    test_real_distribution<normal_distribution>();
    test_real_distribution<lognormal_distribution>();
    test_real_distribution<chi_squared_distribution>();
    test_real_distribution<cauchy_distribution>();
    test_real_distribution<fisher_f_distribution>();
    test_real_distribution<student_t_distribution>();
    test_distribution(discrete_distribution<unsigned short>({1, 2, 3, 4, 5}));
    test_distribution(discrete_distribution<unsigned int>({1, 2, 3, 4, 5}));
    test_distribution(discrete_distribution<unsigned long>({1, 2, 3, 4, 5}));
    test_distribution(discrete_distribution<unsigned long long>({1, 2, 3, 4, 5}));
    test_distribution(discrete_distribution<short>({1, 2, 3, 4, 5}));
    test_distribution(discrete_distribution<int>({1, 2, 3, 4, 5}));
    test_distribution(discrete_distribution<long>({1, 2, 3, 4, 5}));
    test_distribution(discrete_distribution<long long>({1, 2, 3, 4, 5}));

    vector<float> rangesf{0.0f, 10.0f, 90.0f, 100.0f};
    vector<float> weightsf{1.0f, 0.0f, 0.0f, 1.0f};
    test_distribution(piecewise_constant_distribution<float>(begin(rangesf), end(rangesf), begin(weightsf)));
    test_distribution(piecewise_linear_distribution<float>(begin(rangesf), end(rangesf), begin(weightsf)));

    vector<double> ranges{0.0, 10.0, 90.0, 100.0};
    vector<double> weights{1.0, 0.0, 0.0, 1.0};
    test_distribution(piecewise_constant_distribution<double>(begin(ranges), end(ranges), begin(weights)));
    test_distribution(piecewise_linear_distribution<double>(begin(ranges), end(ranges), begin(weights)));

    vector<long double> rangesl{0.0l, 10.0l, 90.0l, 100.0l};
    vector<long double> weightsl{1.0l, 0.0l, 0.0l, 1.0l};
    test_distribution(piecewise_constant_distribution<long double>(begin(rangesl), end(rangesl), begin(weightsl)));
    test_distribution(piecewise_linear_distribution<long double>(begin(rangesl), end(rangesl), begin(weightsl)));
}
```

chris0x44 commented:
Thanks for the test-results and the samples. I looked into them and will upload a fix shortly.

The first failure was caused by the missing static_cast<_Diff> for the return value (as it was already in the original version).

The second issue was the assumption that `_Bitrange` can't exceed the number of bits in the return-type. Considering that `_Value` discards exceeding bits, `_Bitrange` (the number of bits generated) also has to be restricted to _Udiff_bits.

chris0x44 commented on stl/inc/xutility:
```diff
@@ -3,6 +3,34 @@
 
+
+// Copyright 2018 Ulf Adams
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
For the time being, I'll remove the redundant line, with the next change.

BillyONeal commented:
This is still failing internal tests right now due to missing casts, will try to add those and let you know.

BillyONeal commented:
@chris0x44 Can you double check that you are happy with the changes I pushed here?

chris0x44 commented:
@BillyONeal The changes look fine, thanks for fixing the issues. :)
By the way, is there something I can adapt in my setup? Just building from VS2019 (via File -> Open CMake then build) was silent for me.

BillyONeal commented:
In this case I don't think there's anything you could have done because our tests aren't here yet, and the build break didn't appear until doing a merge (your most recent commit passed building here).

BillyONeal commented:
Deassigning myself because even though I verified that the current state passes tests, Stephan still has comments.

statementreply commented on stl/inc/xutility:
```diff
@@ -4823,24 +4852,130 @@ public:
+        for (;;) { // try a sample random value
+            // _Ret is [0, _Mask], _UIndex - 1 <= _Mask, return if unbiased
+            if (_Ret < _Threshold || _Is_power_of2(_UIndex)) {
+                return static_cast<_Diff>(_Reduce(_Ret, _UIndex, _Bitrange));
```
This rejection criteria is incorrect for multiply-shift.

Assume that we'd like to generate a uniform random integer in [0, _OutRange_) with a uniform random integer generator in [0, _InRange_), and that _OutRange_ <= _InRange_.

Let _MapQuo_ = intdiv(_InRange_, _OutRange_), _MapRem_ = mod(_InRange_, _OutRange_). Ideally, we'd like to map _MapQuo_ input values to each output value to make the output uniform. A total of _MapQuo_ * _OutRange_ = _InRange_ - _MapRem_ input values are accepted, and the remaining _MapRem_ are rejected.

Let _X_ be an integer in [0, _InRange_), _Y_ = _X_ * _OutRange_, _YQuo_ = intdiv(_Y_, _InRange_), _YRem_ = mod(_Y_, _InRange_). Thus _YQuo_ ∈ [0, _OutRange_), _YRem_ ∈ [0, _InRange_).

If we accept _YRem_ >= _MapRem_ and reject _YRem_ < _MapRem_, there will be exactly _MapQuo_ input values _X'_ producing _YQuo'_ = _YQuo_, _YRem'_ ∈ [_MapRem_, _InRange_) and _YRem'_ ≡ _YRem_ (mod _OutRange_).

Possible implementation:
```
const auto _Out_max = static_cast<_Udiff>(_Index - 1);

// multiply-shift
_Udiff _Hi = /* (_Ret * _UIndex) >> _Bitrange */;
_Udiff _Lo = /* (_Ret * _UIndex) & _Mask */;

// save a division when _Lo >= _Index - 1 >= _Threshold
if (_Lo < _Out_max) {
    return static_cast<_Diff>(_Hi);
}

// rejection threshold
const auto _Threshold = static_cast<_Udiff>((_Mask - _Out_max) % _UIndex);
for (;;) {
    if (_Lo >= _Threshold) {
        return static_cast<_Diff>(_Hi);
    }
    _Ret = /* get _Bitrange random bits */;
    _Hi = /* (_Ret * _UIndex) >> _Bitrange */;
    _Lo = /* (_Ret * _UIndex) & _Mask */;
}
```

chris0x44 replied:
You're right. Thanks for pointing it out and the additional explanation. I will change it accordingly.

statementreply commented on stl/inc/xutility:
```diff
@@ -4823,24 +4852,130 @@ public:
+            if (_Ret < _Threshold || _Is_power_of2(_UIndex)) {
+                return static_cast<_Diff>(_Reduce(_Ret, _UIndex, _Bitrange));
             }
+            // re-use some bits of the previous iteration
```
Bits of the previous iteration are no longer uniformly distributed given the condition that they are rejected. _Bitrange new bits need to be generated for each iteration.

chris0x44 replied:
Will be changed.

chris0x44 commented:
Fixed issues found by @statementreply also added calculation of _Bits and _Bmask at compiletime

Here are the new measurements:

> Run on (8 X 2592 MHz CPU s)
> CPU Caches:
> L1 Data 32K (x4)
> L1 Instruction 32K (x4)
> L2 Unified 262K (x4)
> L3 Unified 6291K (x1)

**Engine: std::mt19937**
Target: x86

```
-------------------------------------------------------
Benchmark             Time             CPU   Iterations
-------------------------------------------------------
baseline_64        37.9 ns         38.5 ns     19478261
optimized_64       21.2 ns         21.3 ns     37333333

baseline_32        15.3 ns         15.4 ns     49777778
optimized_32       8.23 ns         8.16 ns     74666667

baseline_16        15.0 ns         15.1 ns     49777778
optimized_16       9.07 ns         9.00 ns     74666667

baseline_8         17.0 ns         17.1 ns     44800000
optimized_8        10.3 ns         10.3 ns     64000000
```

Target: x64

```
-------------------------------------------------------
Benchmark             Time             CPU   Iterations
-------------------------------------------------------
baseline_64        44.3 ns         44.5 ns     17920000
optimized_64       18.6 ns         18.8 ns     40727273

baseline_32        12.9 ns         12.9 ns     44800000
optimized_32       6.57 ns         6.56 ns    112000000

baseline_16        13.3 ns         13.4 ns     56000000
optimized_16       7.22 ns         7.25 ns    112000000

baseline_8         13.3 ns         13.4 ns     56000000
optimized_8        7.89 ns         7.85 ns     89600000
```

**Engine: std::ranlux24_base**
Target: x86

```
baseline_64        62.9 ns         62.8 ns     11200000
optimized_64       38.1 ns         37.7 ns     22400000

baseline_32        24.7 ns         25.1 ns     24888889
optimized_32       20.0 ns         20.1 ns     37333333

baseline_16        18.9 ns         19.0 ns     34461538
optimized_16       20.2 ns         20.1 ns     37333333

baseline_8         24.8 ns         24.6 ns     29866667
optimized_8        18.7 ns         18.8 ns     37333333
```

Target: x64

```
baseline_64        41.3 ns         41.4 ns     16592593
optimized_64       17.1 ns         17.3 ns     40727273

baseline_32        22.0 ns         22.0 ns     32000000
optimized_32       18.8 ns         18.6 ns     34461538

baseline_16        17.9 ns         17.6 ns     37333333
optimized_16       12.3 ns         12.3 ns     56000000

baseline_8         15.2 ns         15.4 ns     49777778
optimized_8        12.6 ns         12.6 ns     56000000
```

StephanTLavavej commented:
Marked as `decision needed` because if this changes the runtime behavior of the distribution, we need to decide whether that's acceptable in a binary-compatible world. (Apologies for taking so long with this PR, now the oldest non-draft PR.)

chris0x44 commented:
@StephanTLavavej No worries.
With regard to the change, there was the question, whether the umul and shiftright code copied from ryu should be extracted into a separate header. I guess this will be done in a separate step? Are there any further changes you need from my side?

StephanTLavavej commented:
> there was the question, whether the umul and shiftright code copied from ryu should be extracted into a separate header. I guess this will be done in a separate step?

Yeah - I need to think about how to structure this for future Ryu updates.

> Are there any further changes you need from my side?

Not at this moment - thanks for asking.

Currently I am reviewing/revising the make_shared for arrays PR in order to unblock our new intern's work, then I have a backlog of changelog entries to process, then I should be able to look at this PR again.

cbezault commented:
/azp run

cbezault commented:
/azp run

cbezault commented:
/azp run

cbezault commented:
/azp run

lemire commented:
Not that it is very important, but you guys might prefer to refer to the formal scientific reference rather than to a blog post and a preprint...

Daniel Lemire, Fast Random Integer Generation in an Interval, ACM Transactions on Modeling and Computer Simulation 29 (1), 2019.


Cheers!

chris0x44 commented:
@lemire Thanks for the hint. I updated the comment in the code to use your reference.

lemire commented:
@chris0x44 The GNU libstdc++ library has a faster uniform_int_distribution as of yesterday:

https://gcc.gnu.org/git/?p=gcc.git;a=blobdiff;f=libstdc%2B%2B-v3/include/bits/uniform_int_dist.h;h=ecb8574864aee10b9ea164379fffef27c7bdb0df;hp=6e1e3d5fc5fe8f7f22e62a85b35dc8bfa4743372;hb=98c37d3bacbb2f8bbbe56ed53a9547d3be01b66b;hpb=6ce2cb116af6e0965ff0dd69e7fd1925cf5dc68c

cbezault commented:
@lemire, if only we could look at that XD

cbezault commented:
/azp run

lemire commented:
Ah. Possibly you don’t want to look at it because of licensing issues. If so, sorry, I did not mean to cause harm.

cbezault commented:
/azp run

MattStephanson commented:
Hi @chris0x44. Are you still interested/able to work on this? I ask because I'd like to see it implemented as a prelude to investigating https://hal.archives-ouvertes.fr/hal-03282794/document as a solution to #1074. I think I understand Lemire's method well enough to either implement it myself or land this PR, but I wanted to give you the opportunity to finish it yourself if you still want to.

chris0x44 commented:
Hi @MattStephanson, thanks fpr pinging me on this. To be perfectly honest, I sadly don't have the capacity to work on this. Which becomes painfully evident by me spacing out that there were pending issues on this PR. If you'd like to take this over, feel free. Sorry for causing such delay.

lemire commented:
I am available if someone needs a coder to work on this, or to help with the implementation. Just ping me. I am also available to help with https://hal.archives-ouvertes.fr/hal-03282794/document (I know the author.)

MattStephanson commented:
@chris0x44 No problem, thanks for taking the time to respond.

@lemire I don't mind working on it (unless you're saying that you prefer to make a PR yourself). I have a basic implementation working, I just want to consult with the maintains about which of the several full-width 64x64 bit products I can use, and how much can be `constexpr` without breaking ABI. Particularly when we get to the uniform real distribution, I expect we'll have to precalculate as much as possible to get competitive performance.

lemire commented:
@MattStephanson 

> Particularly when we get to the uniform real distribution, I expect we'll have to precalculate as much as possible to get competitive performance.

I agree that there are interesting challenges. That's why I am offering my help.
